Spack recipes for AOCL 3.0

amd-toolchain-support opened the pull request:
(1) blis - changed configure function with configure_args name which
    will return args to configure function

(2) amdblis - reused configure_args from blis package using super
    function and updated the amdblis specific flags for 3.0 version

(3) amdlibflame - Enabling '--enable-f2c-dotc' flags from version 3.0
    onwards

(4) amdlibm - Added reference to 3.0 release branch

(5) amdscalapack - Added reference to 3.0 release

(6) aocl-sparse - Added suport for ilp64 variant from version 3.0
   Incorporated review comments from previous below pull request
   https://github.com/spack/spack/pull/22276

amd-toolchain-support committed:
Spack recipes for AOCL 3.0 recipe

(1) blis - changed configure function with configure_args name which
    will return args to configure function

(2) amdblis - reused configure_args from blis package using super
    function and updated the amdblis specific flags for 3.0 version

(3) amdlibflame - Enabling '--enable-f2c-dotc' flags from version 3.0
    onwards

(4) amdlibm - Added reference to 3.0 release branch

(5) amdscalapack - Added reference to 3.0 release

(6) aocl-sparse - Added suport for ilp64 variant from version 3.0
   Incorporated review comments from previous below pull request
   https://github.com/spack/spack/pull/22276

tldahlgren commented on var/spack/repos/builtin/packages/amdscalapack/package.py:
```diff
@@ -38,11 +39,13 @@ def cmake_args(self):
-                'ON' if '%aocc ^amdblis' in spec else 'OFF'
-            )
-        ])
+        if "%gcc@10:" in spec:
```
```suggestion
        if spec.satisfies('%gcc@10:'):
```

vkallesh replied:
Incorporated review comments

tldahlgren commented on var/spack/repos/builtin/packages/amdscalapack/package.py:
```diff
@@ -38,11 +39,13 @@ def cmake_args(self):
+            args.extend(['-DCMAKE_Fortran_FLAGS={0}'.format(
+                        "-fallow-argument-mismatch")])
+
+        if "@2.2" in spec:
```
```suggestion
        if spec.satisfies('@2.2'):
```

vkallesh replied:
Incorporated review comments

tldahlgren commented on var/spack/repos/builtin/packages/amdlibflame/package.py:
```diff
@@ -56,6 +55,12 @@ def configure_args(self):
+
+        """To enabled Fortran to C calling convention for
+        complex types when compiling with aocc flang"""
+        if "@3.0: %aocc" in self.spec:
```
```suggestion
        if self.spec.satisfies('@3.0: %aocc'):
```

vkallesh replied:
Incorporated review comments

tldahlgren commented on var/spack/repos/builtin/packages/amdblis/package.py:
```diff
@@ -17,9 +17,21 @@ class Amdblis(BlisBase):
+        spec = self.spec
+        args = super(Amdblis, self).configure_args()
+
+        if "@3.0%aocc@3.0.0" in spec:
```
```suggestion
        if spec.satisfies('@3.0%aocc@3.0.0'):
```

vkallesh replied:
Incorporated review comments

tldahlgren commented on var/spack/repos/builtin/packages/amdscalapack/package.py:
```diff
@@ -38,11 +39,13 @@ def cmake_args(self):
+
+        if "@2.2" in spec:
+            args.extend(['-DUSE_DOTC_WRAPPER:BOOL=%s' % (
+                        'ON' if '%aocc ^amdblis' in spec else 'OFF')])
```
```suggestion
                        'ON' if spec.satisfies('%aocc ^amdblis') else 'OFF')])
```

vkallesh replied:
Incorporated review comments

tldahlgren commented on var/spack/repos/builtin/packages/aocl-sparse/package.py:
```diff
@@ -64,18 +67,19 @@ def cmake(self, spec, prefix):
 
-        with working_dir(self.build_directory, create=True):
-            inspect.getmodule(self).cmake(*args)
+        if "@3.0:" in spec:
```
```suggestion
        if spec.satisfies('@3.0:'):
```

tldahlgren commented on var/spack/repos/builtin/packages/aocl-sparse/package.py:
```diff
@@ -64,18 +67,19 @@ def cmake(self, spec, prefix):
+        return args
 
     # Check that self.prefix is there after installation
     @run_after('build')
```
FWIW. Regarding the comment above, `check` is technically testing the build -- after the `build` phase and using the `self.build_directory` -- not the installation.  Testing the installation would be done after the `install` phase and would be using the `self.prefix`, which is the root directory for the package's installed files.

vkallesh replied:
Removed following comment:
# Check that self.prefix is there after installation

Changed the following text from
Simple test to test the installation by running
to
Simple test to test the built library by running

tldahlgren reviewed (changes requested):
Made a number of minor suggested changes and comments.  

We provide more examples of checking a spec's constraints at https://spack.readthedocs.io/en/latest/packaging_guide.html?highlight=spec.satisfies#testing-spec-constraints.

amd-toolchain-support committed:
Below changes were done for AOCL 3.0 recipes:

 (1) blis - Due to super function in amdblis, config_arg 'auto'
	 comes in the middle of configure command argument, resulting
	 in build failure for amdblis with aocc compiler.
         To avoid this, under blis spack recipe moved config_arg 'auto'
	 from configure_args to configure function.
	 This change will ensure auto should always be the last argument for
	 configure command for both base (blis)and derived class (amdblis)

 (2) amdblis - Incorporated review comments

 (3) amdscalapack - Incorporated review comments

 (4) amdscalapack - Incorporated review comments

amd-toolchain-support commented:
> Made a number of minor suggested changes and comments.
> 
> We provide more examples of checking a spec's constraints at https://spack.readthedocs.io/en/latest/packaging_guide.html?highlight=spec.satisfies#testing-spec-constraints.

Thanks for the suggestion, all review comments are addressed.

tldahlgren commented on var/spack/repos/builtin/packages/amdblis/package.py:
```diff
@@ -17,9 +17,22 @@ class Amdblis(BlisBase):
+        spec = self.spec
+        args = super(Amdblis, self).configure_args()
+
+        if spec.satisfies('@3.0%aocc'):
```
```suggestion
        if spec.satisfies('@3.0: %aocc'):
```

tldahlgren replied:
The above assumes the code in the body of the `if` statement applies to versions >= `3.0`.

amd-toolchain-support replied:
These flags are specific to amdblis 3.0 version. Next version of amdblis don't need below flags. 

 if spec.satisfies('@3.0 %aocc'):
            """ To enabled Fortran to C calling convention for
            complex types when compiling with aocc flang"""
            args.append('CFLAGS={0}'.format("-DAOCL_F2C"))
            args.append('CXXFLAGS={0}'.format("-DAOCL_F2C"))

tldahlgren commented on var/spack/repos/builtin/packages/amdscalapack/package.py:
```diff
@@ -38,11 +39,13 @@ def cmake_args(self):
+            args.extend(['-DCMAKE_Fortran_FLAGS={0}'.format(
+                        "-fallow-argument-mismatch")])
+
+        if spec.satisfies('@2.2'):
```
Does this argument only apply to version `2.2`?  If not, you might want to use a range check here.  

https://spack.readthedocs.io/en/latest/packaging_guide.html?highlight=spec.satisfies#testing-spec-constraints

amd-toolchain-support replied:
Yes, this argument only apply to 2.2 version.

tldahlgren commented on var/spack/repos/builtin/packages/aocl-sparse/package.py:
```diff
@@ -29,8 +29,10 @@ class AoclSparse(CMakePackage):
+            description='Build with ILP64 support')
 
-    depends_on('boost')
+    depends_on('boost', when='@2.2')
```
Again, does this dependency only apply to version `2.2`.

amd-toolchain-support replied:
Yes, Boost is required only for aocl-sparse 2.2 version.

tldahlgren commented on var/spack/repos/builtin/packages/amdlibm/package.py:
```diff
@@ -37,8 +35,8 @@ class Amdlibm(SConsPackage):
 
-    patch('0001-libm-ose-Scripts-cleanup-pyc-files.patch')
-    patch('0002-libm-ose-prevent-log-v3.c-from-building.patch')
+    patch("0001-libm-ose-Scripts-cleanup-pyc-files.patch", when="@2.2")
```
Is there a version before `2.2` that could be installed?  If so, then you'd want to use `@:2.2`.

amd-toolchain-support replied:
amdlibm 2.2 is the first version in use for spack. There is no previous version.

amd-toolchain-support committed:
Introduced spack between amdblis and compiler version

tldahlgren reviewed (approved):
LGTM.  Approval contingent on all required checks passing.

Thanks for responding to my questions.

tldahlgren commented:
Thank you for your contribution!
